[Shipping labels] Label refund button visibility logic

irfano opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->
Closes WOOMOB-370

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds additional logic for the visibility of the “Request refund” button. The logic was copied from the old flow.

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
Open purchased shipments. For shipments purchased more than 30 days ago, the “Request refund” button should not be displayed.
I couldn't find a way to test ANONYMIZED labels, but I copied the condition from the old flow, so it should be safe.

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
Steps above.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
|Before|After|
|-|-|
|<img src="https://github.com/user-attachments/assets/ec2af655-2d86-4011-89f7-106cf1e3633d" width=340>|<img src="https://github.com/user-attachments/assets/a484271c-43b9-4977-b337-f074a251222f" width=340>|

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

irfano committed:
Expose full ShippingLabelModel in ObserveShippingLabelStatus

This change modifies `ObserveShippingLabelStatusResult` to include the entire `ShippingLabelModel` instead of just the `refundableAmount`.

irfano committed:
Hide refund button when a refund is not available

irfano committed:
Encapsulate label details into ShippingLabelModel

This commit refactors the `ShipmentUIModel` to encapsulate shipping label-related information (ID, carrier ID, tracking number, status, etc.) into a new `ShippingLabelModel`.

irfano committed:
Set refund expiry days to 30

irfano committed:
Disable refund if label has expired

Implement the refund expiry for a shipping label, so if it's older than 30 days old, it's not possible to request a refund for it.
Also, if the label status is `ANONYMIZED` it's not possible to request a refund either.

irfano committed:
Fix GetShipmentsTests

irfano committed:
Fix WooShippingLabelCreationViewModelTest

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/models/ShippingLabelModel.kt:
```diff
@@ -25,7 +29,18 @@ data class ShippingLabelModel(
+) : Parcelable {
+    val isRefundAvailable: Boolean
+        get() {
+            val createdTime = createdDate?.time ?: return true
+            return status != ShippingLabelStatus.ANONYMIZED &&
+                Date().before(Date(createdTime + TimeUnit.DAYS.toMillis(REFUND_EXPIRY_DAYS))) == true
```
This logic was copied from [ShippingLabelRefundViewModel](https://github.com/woocommerce/woocommerce-android/blob/0d26c454ff1b17d55bae755ad1091a20d4094e38/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/ShippingLabelRefundViewModel.kt#L82)

hichamboushaba replied:
np, I know I'm the one who wrote the logic in the old flow, but now that I'm thinking about it, I think using `Date` for this is kinda an overkill, WDYT about something simpler like:

```suggestion
            val createdTime = createdDate?.time ?: return true
            return status != ShippingLabelStatus.ANONYMIZED &&
                System.currentTimeMillis() < createdTime + TimeUnit.DAYS.toMillis(REFUND_EXPIRY_DAYS)
```

The above avoids two instantiations, and to me the readability is not impacted at all.

irfano replied:
Updated: 09141d9072beea704c422452f4ced2d7b860ada7
Also added test for `isRefundAvailable`

hichamboushaba reviewed (dismissed):
Good work @irfano, I left just on np comment, and I have a suggestion: I think we should have some testing for the `isRefundAvailable` logic, WDYT?

Otherwise, good to go and I'm pre-approving.

hichamboushaba commented:
@irfano while testing the iOS app, I noticed that the expiration duration for the refund is not a constant value, it depends on the carriers (we also mention this in the M4 testing plan), and it's returned from the API (the field `expiry_date`), so what do you prefer: merge this PR then update its logic on a separate PR, or do you want me to dismiss my review and you work on the new logic?

For reference, here is the iOS [PR](https://github.com/woocommerce/woocommerce-ios/pull/15603)

hichamboushaba commented:
Another topic about this, we seem to be [hardcoding](https://github.com/woocommerce/woocommerce-android/blob/3242df4b93fbccb4fc39922d2d80b3d91b27ceeb/WooCommerce/src/main/res/values/strings.xml#L4671) the "14 days" duration for the refund in the refund screen, while it should depend on the carrier ID ([here](https://github.com/woocommerce/woocommerce-android/blob/3242df4b93fbccb4fc39922d2d80b3d91b27ceeb/WooCommerce/src/main/res/values/strings.xml#L4671) in the plugin), do you want to handle the same change on the same PR or do you prefer me to create an issue about this?

irfano commented:
> so what do you prefer: merge this PR then update its logic on a separate PR, or do you want me to dismiss my review and you work on the new logic?

Since it’s related to the refund logic, I’d prefer to address it in this PR. I missed the backend logic in the M4 test plan, so I’ll include it here. You can dismiss your review for now, I’ll request it again once I’m done.

irfano committed:
[Shipping Labels] Enhance refund availability logic

irfano committed:
Update refund duration text based on shipping carrier

The refund description text on the refund screen is updated to show the correct refund duration based on the shipping carrier. DHL Express refunds are processed in 14 days, while other carriers take 31 days.

irfano commented:
O copied the backend logic for isRefundAvailable and its tests. Also updated the 14/31 days logic in this PR. This is ready for another review, @hichamboushaba 
I couldn't test 14-day case for the Dhl Express carrier, as I couldn't find any DHL Express rates on either apps or the web.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/models/ShippingLabelModel.kt:
```diff
@@ -29,16 +30,35 @@ data class ShippingLabelModel(
+    val usedDate: Long?,
 ) : Parcelable {
+    // Following logic in the plugin
+    // https://github.com/woocommerce/woocommerce-shipping/blob/0f67a1eb349cbe90ce471d88c7b31bd4950d6744/client/utils/label/refund.ts
```
This is a private repo, let's not mention it here.

IMO it would be safer to rebase interactively and modify the commit to exclude this change.

irfano replied:
Pinging @itsmeichigo and @RafaelKayumov , since this was also added in [iOS](https://github.com/woocommerce/woocommerce-ios/blob/af68d940e9cf44fe15f23c9b31df9036a23f68ff/WooCommerce/Classes/Extensions/ShippingLabel%2BHelpers.swift).

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/models/ShippingLabelModel.kt:
```diff
@@ -29,16 +30,35 @@ data class ShippingLabelModel(
+
     companion object {
         private const val REFUND_EXPIRY_DAYS = 30L
+        private const val REFUND_DURATION_DHL_EXPRESS = 14
+        const val REFUND_DURATION_DEFAULT = 31
```
I believe these are inverted, it should be 31 for `dhlexpress` and 14 for the others, can you double check please?

irfano replied:
Good catch! Fixed: 60b31d7dc13ac059d1f3c80400ee8f01aad4acf5

hichamboushaba reviewed (commented):
Nice work @irfano, the code looks good, but there are two issues that should be addressed before merging this, please check my comments.

irfano committed:
Fix refund expiration dates

The refund expiration dates for DHL Express and other carriers were swapped. This commit corrects them.

hichamboushaba reviewed (approved):
Nice work, sorry I missed re-reviewing this after the last changes.
